### PR TITLE
Pageview tracking

### PR DESCRIPTION
**Pull request by jaydp17**
**Status**
- [x] WIP
- [ ] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api
- hyperion (frontend)

**Run database migrations (delete if no migration was added)**
YES

**Release notes for users (delete if codebase-only change)**
- Tracks PageViews (will come back and edit this)

**What all is done?**
- Data is stored in a tabled called `pageviews`
- Mutation to record community, channel & thread view event along with the `refererDomain`
- Query/Resolvers to get community, channel & thread event data grouped by `date` as well as `refererDomain`

**What's pending?**
- Query for popular content
- Date filter in queries
- Calling the mutation/query from the frontend

### CODE REVIEW

File: api/migrations/20180503065927-pageview-table.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+exports.up = function(r, conn) {
+  return r
+    .tableCreate('pageview')
```
Comment (mxstbr) [inline api/migrations/20180503065927-pageview-table.js]: Let's call the table in plural `"pageviews"` like our other tables

File: api/types/Community.js [hunk 2]
```diff
@@ -272,6 +272,11 @@ const Community = /* GraphQL */ `
 		id: ID!
 		customMessage: String
 	}
+	
+	input AddPageViewInput {
+		id: String!
+		referrerDomain: String
```
Comment (mxstbr) [inline api/types/Community.js]: Idea: to support other types of pageviews we could add an enum `referrerType` argument here:

```JS
enum ReferrerType {
  COMMUNITY
  THREAD
  CHANNEL
}

input AddPageViewInput {
  id: String!
  referrerDomain: String
  referrerType: ReferrerType!
}
```

WDYT?

**Review by mxstbr — COMMENTED** [review]
Looking great so far! 😍

File: api/types/Community.js (hunk 2 above)
Comment (jaydp17) [reply api/types/Community.js]: yes, it makes sense.
Then we'll move it out of community, to a separate directory, as it's no more just about community

Comment (mxstbr) [reply api/types/Community.js]: Sounds good to me

File: api/mutations/analytics/addPageView.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+// @flow
+
+import { parse as parseURL } from 'url';
```
Comment (mxstbr) [inline api/mutations/analytics/addPageView.js]: Minor nitpick: `url.parse` is deprecated, it's recommended to use the `URL` constructor now: https://nodejs.org/api/url.html#url_constructor_new_url_input_base

```JS
import { URL } from 'url';

// ...

const parsedUrl = new URL(url);
```

Comment (jaydp17) [reply api/mutations/analytics/addPageView.js]: oh, didn't know that it was deprecated.
Thanks for that, I've updated the code.

File: api/models/pageviews.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+};
+
+type GetPageViewParam = {
+  id: string,
+  pageviewType: PageViewType,
+};
+export const getPageViews = ({ id, pageviewType }: GetPageViewParam) => {
+  return db
+    .table('pageviews')
+    .filter({ refId: id, refType: pageviewType.toLowerCase() })
```
Comment (mxstbr) [inline api/models/pageviews.js]: RethinkDB-specific performance note to myself: `.filter` is really slow, we should create indexes for these.

Don't worry about that though, I'll take care of it before we ship this!

Comment (jaydp17) [reply api/models/pageviews.js]: @mxstbr does this (https://github.com/withspectrum/spectrum/pull/3005/commits/637b15a170356134127aa94e0b05f5a9ea256bb6) work?

Comment (mxstbr) [reply api/models/pageviews.js]: Yep that totally works! 😍

**Pull request closed without merging**